OrbitGl/TrackManager.cpp

```diff
@@ -0,0 +1,449 @@
+      auto& previous_filtered_track = sorted_filtered_tracks_[moving_track_current_position - 1];
+      sorted_tracks_.insert(
+          ++std::find(sorted_tracks_.begin(), sorted_tracks_.end(), previous_filtered_track),
+          moving_track);
+    } else {
+      // In this case we will insert the moving_track right before the one who is after in the
+      // filtered array
+      auto& next_filtered_track = sorted_filtered_tracks_[moving_track_current_position + 1];
+      sorted_tracks_.insert(
+          std::find(sorted_tracks_.begin(), sorted_tracks_.end(), next_filtered_track),
+          moving_track);
+    }
+  }
+}
+
+int TrackManager::FindMovingTrackIndex() {
+  // Returns the position of the moving track, or -1 if there is none.
+  for (auto track_it = sorted_filtered_tracks_.begin(); track_it != sorted_filtered_tracks_.end();
+       ++track_it) {
+    if ((*track_it)->IsMoving()) {
+      return track_it - sorted_filtered_tracks_.begin();
+    }
+  }
+  return -1;
+}
+
+void TrackManager::UpdateTracks(uint64_t min_tick, uint64_t max_tick, PickingMode picking_mode) {
+  // Make sure track tab fits in the viewport.
+  TimeGraphLayout layout = time_graph_->GetLayout();
+
+  float current_y = -layout.GetSchedulerTrackOffset() - layout.GetTrackTabHeight();
+  float pinned_tracks_height = 0.f;
+
+  // Draw pinned tracks
+  for (auto& track : sorted_filtered_tracks_) {
+    if (!track->IsPinned()) {
+      continue;
+    }
+
+    const float z_offset = GlCanvas::kZOffsetPinnedTrack;
+    track->SetY(current_y + time_graph_->GetCanvas()->GetWorldTopLeftY() - layout.GetTopMargin() -
+                layout.GetSchedulerTrackOffset());
+    track->UpdatePrimitives(min_tick, max_tick, picking_mode, z_offset);
+    const float height = (track->GetHeight() + layout.GetSpaceBetweenTracks());
+    current_y -= height;
+    pinned_tracks_height += height;
+  }
+
+  // Draw unpinned tracks
+  for (auto& track : sorted_filtered_tracks_) {
+    if (track->IsPinned()) {
+      continue;
+    }
+
+    const float z_offset = track->IsMoving() ? GlCanvas::kZOffsetMovingTack : 0.f;
+    track->SetY(current_y);
+    track->UpdatePrimitives(min_tick, max_tick, picking_mode, z_offset);
+    current_y -= (track->GetHeight() + layout.GetSpaceBetweenTracks());
+  }
+
+  tracks_height_ = current_y;
+}
+
+void TrackManager::AddTrack(std::shared_ptr<Track> track) {
+  tracks_.emplace_back(track);
+  sorting_invalidated_ = true;
+}
+
+void TrackManager::RemoveFrameTrack(uint64_t function_address) {
+  frame_tracks_.erase(function_address);
+  sorting_invalidated_ = true;
+}
+
+SchedulerTrack* TrackManager::GetOrCreateSchedulerTrack() {
+  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  std::shared_ptr<SchedulerTrack> track = scheduler_track_;
+  if (track == nullptr) {
+    track = std::make_shared<SchedulerTrack>(time_graph_);
+    AddTrack(track);
+    scheduler_track_ = track;
+    uint32_t num_cores = time_graph_->GetNumCores();
```

Comment:
Consider using the number of cores from the layout.

Reply:
b/176086740